include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -306,6 +306,82 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
   }
 }
 
+
+
+template<int dim, int spacedim>
+std::vector<std::shared_ptr<const Triangulation<dim, spacedim>>>
+create_geometric_coarsening_sequence(const Triangulation<dim, spacedim> & fine_triangulation_in)
+{
+  std::vector<std::shared_ptr<const Triangulation<dim, spacedim>>> coarse_grid_triangulations(
+    fine_triangulation_in.n_global_levels());
+
+  coarse_grid_triangulations.back().reset(&fine_triangulation_in, [](auto *) {
+    // empty deleter, since fine_triangulation_in is an external field
+    // and its destructor is called somewhere else
+  });
+
+  // for a single level nothing has to be done
+  if(fine_triangulation_in.n_global_levels() == 1)
```

Comment:
Generally, I would prefer

```suggestion
  if(fine_triangulation_in.n_global_levels() > 1)
  {
   ...
  }
  
  return coarse_grid_triangulations;
```

instead of several return statements within one function. This also renders the comment in line 323 superfluous.